src/typecheck/mod.rs

```diff
@@ -1381,51 +1435,88 @@ fn type_check_<L: Linearizer>(
-                        // and we need to use the same.
-                        let ty = if let Term::RecRecord(..) = t.as_ref() {
-                            ctxt.type_env.get(id).cloned().unwrap()
-                        } else {
-                            state.table.fresh_type_uvar()
-                        };
-
-                        type_check_field(
+                // Building the type {id1 : ?a1, id2: ?a2, .., idn: ?an}
+                let mut rows_expected_type: HashMap<Ident, UnifType> = record
+                    .fields
+                    .keys()
+                    .map(|id| (*id, state.table.fresh_type_uvar()))
+                    .collect();
+
+                let rows_skeleton =
+                    rows_expected_type
+                        .iter()
+                        .fold(mk_uty_row!(), |acc, (id, row_ty)| {
+                            // if let Term::RecRecord(..) = t.as_ref() {
+                            //     ctxt.type_env.get(id).cloned().unwrap()
+                            // }
+                            mk_uty_row!((*id, row_ty.clone()); acc)
+                        });
+
+                unify(state, &ctxt, ty, mk_uty_record!(; rows_skeleton))
+                    .map_err(|err| err.into_typecheck_err(state, rt.pos))?;
+
+                for (id, field) in record.fields.iter() {
+                    if let Term::RecRecord(..) = t.as_ref() {
+                        let affected_type = ctxt.type_env.get(id).cloned().unwrap();
+                        unify(
                             state,
-                            ctxt.clone(),
-                            lin,
-                            linearizer.scope(),
-                            *id,
-                            field,
-                            ty.clone(),
-                        )?;
+                            &ctxt,
+                            rows_expected_type.get(id).cloned().unwrap(),
+                            affected_type,
+                        )
+                        .map_err(|err| {
+                            err.into_typecheck_err(
+                                state,
+                                field.value.as_ref().map(|v| v.pos).unwrap_or_default(),
+                            )
+                        })?;
+                    }
 
-                        Ok(mk_uty_row!((*id, ty); acc))
-                    },
-                )?;
+                    check_field(
+                        state,
+                        ctxt.clone(),
+                        lin,
+                        linearizer.scope(),
+                        *id,
+                        field,
+                        // unwrap(): we've built `rows_expected_type` in this very function
+                        // from record.fields.keys(), so it must contain `id`
+                        rows_expected_type.remove(id).unwrap(),
```

Comment:
Nitpick: we could use `expect` here & put the comment into the string argument.

Reply:
Hmm, I always wonder about this dilemma. My usual line of thought is that I don't see a big advantage of doing so:

1. The error message is useless without the code around it: you're not going to extract any useful information from it
2. I don't need it to find the location of the panic, because Rust tracks that for me.

On the other hand, there isn't a big disadvantage of using `expect` either (beside adding a few bytes to the binary). What's your personal rationale for using `expect` ?

Reply:
I tend to view comments as a last-resort method for expressing the purpose of code. My reasoning being that they require manual labour to ensure they're kept in sync with the code they describe. As such, in places where the comment can be merged into the code in some form, my preference is always to do so. 

Here, by using `expect` rather than `unwrap`, we make the justification for `panic`king part of the code that causes the `panic`, which means if that code changes, the justification is more likely to change with it. 

But I marked it as a "nitpick" for a reason: my feelings on this aren't massively strong, and while I'd do it in my own code, I wouldn't insist on it in code written by others. :smile:

Reply:
We could always try using `flux` instead https://github.com/flux-rs/flux :trollface: